framework/src/executioners/Eigenvalue.C

```diff
@@ -61,40 +77,228 @@ Eigenvalue::validParams()
+  _eigen_problem.initialSetup();
+
+#ifdef LIBMESH_HAVE_SLEPC
+
+  // Make sure all PETSc options are setup correctly
+  prepareSolverOptions();
+
+  // Let do an initial solve if a nonlinear eigen solver but not power is used.
+  // The initial solver is a Inverse Power, and it is used to compute a good initial
+  // guess for Newton
+  auto free_power_iterations = getParam<unsigned int>("free_power_iterations");
+  if (free_power_iterations && _eigen_problem.isNonlinearEigenvalueSolver() &&
+      _eigen_problem.solverParams()._eigen_solve_type != Moose::EST_NONLINEAR_POWER)
+  {
+    _eigen_problem.doFreePowerIteration(true);
+    // Set free power iterations
+    setFreeNonlinearPowerIterations(free_power_iterations);
+
+    // Provide vector of ones to solver
+    // "auto_initialization" is on by default and we init the vector values associated
+    // with eigen-variables as ones. If "auto_initialization" is turned off by users,
+    // it is up to users to provide an initial guess. If "auto_initialization" is off
+    // and users does not provide an initial guess, slepc will automatically generate
+    // a random vector as the initial guess. The motivation to offer this option is
+    // that we have to initialize  ONLY eigen variables in multiphysics simulation.
+    if (_eigen_problem.needInitializeEigenVector())
+      _eigen_problem.initEigenvector(1.0);
+
+    _console << " Free power iteration starts" << std::endl;
+
+    // Call solver
+    _eigen_problem.solve();
+
+    // Clear free power iterations
+    clearFreeNonlinearPowerIterations();
+
+    _eigen_problem.doFreePowerIteration(false);
+  }
+#endif
 }
 
 void
 Eigenvalue::execute()
 {
-#if LIBMESH_HAVE_SLEPC
+  // Recovering makes sense for only transient simulations since the solution from
+  // the previous time steps is required.
+  if (_app.isRecovering())
+    return;
+
+  // Several lines of code are needed to make the output system work (copy from "Steady")
+  _time_step = 0;
+  _time = _time_step;
+  _eigen_problem.outputStep(EXEC_INITIAL);
+  _time = _system_time;
+
+  preExecute();
+
+  // The following code of this function is copied from "Steady"
+  // "Eigenvalue" implementation can be considered a one-time-step simulation to
+  // have the code compatible with the rest moose world.
+  _eigen_problem.advanceState();
+
+  // First step in any eigenvalue state solve is always 1 (preserving backwards compatibility)
+  _time_step = 1;
+
+#ifdef LIBMESH_ENABLE_AMR
+
+  // Define the refinement loop
+  unsigned int steps = _eigen_problem.adaptivity().getSteps();
+  for (unsigned int r_step = 0; r_step <= steps; r_step++)
+  {
+#endif // LIBMESH_ENABLE_AMR
+    _eigen_problem.timestepSetup();
+
+    // This loop is for nonlinear multigrids (developed by Alex)
+    for (MooseIndex(_num_grid_steps) grid_step = 0; grid_step <= _num_grid_steps; ++grid_step)
+    {
+
+      // Let us do extra power iterations here if necessary
```

Comment:
From here to LIne 238, I think the code should be in `_eigen_problem.solve()`. We want the `PicardSolve` as the outermost iteration.